File: app/ios/Classes/PreferencesObserverPlugin.swift
import Flutter
import UIKit

public class PreferencesObserverPlugin: NSObject, FlutterPlugin {
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance: PreferencesObserverPlugin = PreferencesObserverPlugin()
        
        // Setup Method Channel for fetching values on-demand
        let methodChannel: FlutterMethodChannel = FlutterMethodChannel(
            name: "com.alloy.alloy_sdk/methods",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        
        // Setup Event Channel for streaming all updates
        let eventChannel: FlutterEventChannel = FlutterEventChannel(
            name: "com.alloy.alloy_sdk/stream",
            binaryMessenger: registrar.messenger()
        )
        let streamHandler: UserDefaultsStreamHandler = UserDefaultsStreamHandler()
        eventChannel.setStreamHandler(streamHandler)
    }
    
    /**
     * Handles incoming method calls from Dart.
     */
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "getValue" {
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Key argument is missing or invalid", details: nil))
                return
            }
            // Retrieve the value from standard UserDefaults for the requested key.
            let value: Any? = UserDefaults.standard.object(forKey: key)
            let serializedValue: Any? = UserDefaultsSerializer.serializeValue(value)
            result(serializedValue)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}

File: app/android/src/main/kotlin/com/alloy/alloy_sdk/PreferencesObserverPlugin.kt
package com.alloy.alloy_sdk

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class PreferencesObserverPlugin: FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var sharedPreferences: SharedPreferences

    companion object {

        const val KEY_PREFIX: String = "IABTCF"
        fun serializeValue(value: Any?): Any? {
            return when (value) {
                null -> null
                is Float -> value.toDouble()
                is Set<*> -> value.filterIsInstance<String>()
                is Boolean, is Int, is Long, is Double, is String -> value
                else -> value.toString()
            }
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        // Use default shared preferences to access keys set by other SDKs like IABTCF_TCString
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

        // Setup Method Channel for fetching values on-demand
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.alloy.alloy_sdk/methods")
        methodChannel.setMethodCallHandler(this)

        // Setup Event Channel for streaming all updates
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.alloy.alloy_sdk/stream")
        val streamHandler = SharedPreferencesStreamHandler(sharedPreferences)
        eventChannel.setStreamHandler(streamHandler)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "getValue") {
            try {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key argument cannot be null", null)
                    return
                }
                // Retrieve the value from SharedPreferences for the requested key.
                val rawValue = sharedPreferences.all[key]
                val value = serializeValue(rawValue)
                result.success(value)
            } catch (e: Exception) {
                result.error("NATIVE_ERROR", "Failed to retrieve value from SharedPreferences", e.toString())
            }
        } else {
            result.notImplemented()
        }
    }
}